google/cloud/pubsub/internal/subscription_session.cc

```diff
@@ -38,38 +39,120 @@ class SubscriptionSessionImpl {
 
-    return self->Start(promise<Status>([self] { self->Shutdown(); }),
-                       std::move(p.callback));
+    auto weak = std::weak_ptr<SubscriptionSessionImpl>(self);
+    auto result = self->shutdown_manager_->Start(promise<Status>([weak] {
+      if (auto self = weak.lock()) self->InitiateApplicationShutdown();
+    }));
+    // Create a (periodic) timer, this serves two purposes:
+    // 1) Effectively the timer owns "self" and the timer is owned by the
+    //    CompletionQueue.
+    // 2) When the completion queue is shutdown, the timer is canceled and
+    //    `self` gets a chance to shutdown the pipeline.
+    self->ScheduleTimer();
+    self->pipeline_->Start(std::move(p.callback));
+    return result.then([weak](future<Status> f) {
+      if (auto self = weak.lock()) self->ShutdownCompleted();
+      return f.get();
+    });
   }
 
   SubscriptionSessionImpl(
+      CompletionQueue cq,
       std::shared_ptr<SessionShutdownManager> shutdown_manager,
-      std::shared_ptr<SubscriptionConcurrencyControl> pipeline)
-      : shutdown_manager_(std::move(shutdown_manager)),
+      std::shared_ptr<SubscriptionConcurrencyControl> pipeline,
+      std::chrono::milliseconds shutdown_polling_period)
+      : cq_(std::move(cq)),
+        shutdown_manager_(std::move(shutdown_manager)),
+        shutdown_polling_period_(shutdown_polling_period),
         pipeline_(std::move(pipeline)) {}
 
-  future<Status> Start(promise<Status> p, pubsub::ApplicationCallback cb) {
-    auto result = shutdown_manager_->Start(std::move(p));
-    pipeline_->Start(std::move(cb));
-    return result;
+ private:
+  void InitiateApplicationShutdown() {
+    GCP_LOG(DEBUG) << __func__ << "()";
+    std::unique_lock<std::mutex> lk(mu_);
+    std::shared_ptr<SubscriptionConcurrencyControl> p = pipeline_;
+    switch (shutdown_state_) {
+      case kNotInShutdown:
+        shutdown_state_ = kShutdownByApplication;
+        lk.unlock();
+        shutdown_manager_->MarkAsShutdown(__func__, {});
+        if (p) p->Shutdown();
+        break;
+      case kShutdownByApplication:
+      case kShutdownByCompletionQueue:
+      case kShutdownCompleted:
+        break;
+    }
   }
 
-  void Shutdown() { pipeline_->Shutdown(); }
+  void ShutdownCompleted() {
+    GCP_LOG(DEBUG) << __func__ << "()";
+    std::lock_guard<std::mutex> lk(mu_);
+    pipeline_.reset();
+    shutdown_state_ = kShutdownCompleted;
+    if (timer_.valid()) timer_.cancel();
+  }
 
- private:
-  std::shared_ptr<SessionShutdownManager> shutdown_manager_;
+  void ScheduleTimer() {
+    GCP_LOG(DEBUG) << __func__ << "()";
+    using TimerArg = future<StatusOr<std::chrono::system_clock::time_point>>;
+
+    std::unique_lock<std::mutex> lk(mu_);
+    auto self = shared_from_this();
+    lk.unlock();
+    auto t = cq_.MakeRelativeTimer(shutdown_polling_period_)
+                 .then([self](TimerArg f) { self->OnTimer(!f.get().ok()); });
+    if (shutdown_state_ == kShutdownByCompletionQueue) return;
```

Comment:
Is `shutdown_state_` not protected by `mu_`?

Reply:
Yikes, fixed, thanks.